infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -0,0 +1,362 @@
+
+        Set<String> newSinks = newTopology.getSinkClusters().keySet();
+        Set<String> currentSinks = topology.getSinkClusters().keySet();
+        Set<String> sinksToRemove = Sets.difference(currentSinks, newSinks);
+        Set<String> sinkClustersUnchanged = Sets.intersection(newSinks, currentSinks);
+
+        topology = newTopology;
+        metadataManager.setTopologyConfigId(topology.getTopologyConfigId());
+
+        Set<LogReplicationSession> sessionsToRemove = new HashSet<>();
+        Set<LogReplicationSession> sessionsUnchanged = new HashSet<>();
+
+        try {
+            IRetry.build(IntervalRetry.class, () -> {
+                try (TxnContext txn = corfuStore.txn(LogReplicationMetadataManager.NAMESPACE)) {
+                    sessions.forEach(session -> {
+                        if (sinksToRemove.contains(session.getSinkClusterId()) ||
+                                sourcesToRemove.contains(session.getSourceClusterId())) {
+                            sessionsToRemove.add(session);
+                            metadataManager.removeSession(txn, session);
+                        }
+
+                        if(sinkClustersUnchanged.contains(session.getSinkClusterId())) {
+                            sessionsUnchanged.add(session);
+                            sessionToContextMap.get(session).setTopologyConfigId(topology.getTopologyConfigId());
+                            metadataManager.updateReplicationMetadataField(txn, session,
+                                    ReplicationMetadata.TOPOLOGYCONFIGID_FIELD_NUMBER, topology.getTopologyConfigId());
+                        }
+                    });
+                    txn.commit();
+                } catch (TransactionAbortedException e) {
+                    throw new RetryNeededException();
+                }
+
+                stopReplication(sessionsToRemove, true);
+                updateReplicationParameters(sessionsUnchanged);
+                createSessions();
+                return null;
+            }).run();
+        } catch (InterruptedException e) {
+            log.error("Unrecoverable exception while removing session", e);
+            throw new UnrecoverableCorfuInterruptedError(e);
+        }
+    }
+
+    private void updateReplicationParameters(Set<LogReplicationSession> sessionsUnchanged) {
+        if (replicationManager == null) {
+            return;
+        }
+        for (LogReplicationSession session : sessionsUnchanged) {
+            ClusterDescriptor cluster = topology.getSinkClusters().get(session.getSinkClusterId());
+            replicationManager.refreshRuntime(session, cluster, topology.getTopologyConfigId());
+        }
+    }
+
+    /**
+     * Create sessions (outgoing and incoming) along with metadata
+     *
+     */
+    private void createSessions() {
+        LogReplicationSession session;
+
+        try(TxnContext txn = corfuStore.txn(LogReplicationMetadataManager.NAMESPACE)) {
```

Comment:
Can we consider wrapping this in a IRetry block too?

Reply:
+1

Reply:
done.